api/jwt.go

```diff
@@ -3,77 +3,227 @@ package api
+//          "verified": false,
+//          "verified_at": null,
+//          "via": "email"
+//        }
+//      ]
+//    },
+//    "issued_at": "2020-12-09T16:09:35.004042Z"
+//  },
+//  "sub": "695725d4-a345-4e68-919a-7395cb68484c"
+//}
 func ValidateToken(t string) (*jwt.Token, error) {
-	token, err := jwt.Parse(t, func(token *jwt.Token) (interface{}, error) {
-		//Make sure that the token method conform to "SigningMethodHMAC"
-		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
-			return nil, errors.New(fmt.Sprintf("unexpected signing method: %v", token.Header["alg"]))
-		}
-		return []byte(os.Getenv("ACCESS_SECRET")), nil
-	})
+	token, err := jwt.Parse(t, keyForToken)
 	if err != nil {
 		return nil, err
 	}
 	if !token.Valid {
 		return nil, errors.New("token is invalid")
 	}
+	// TODO Verify issuer, scope, etc.?
 	return token, nil
 }
 
-// extractToken extracts the JWT token from the request and returns it. Returns
-// an empty string if there is no token.
-func extractToken(r *http.Request) string {
+// keyForToken is a helper function that finds a suitable key for validating the
+// given token among the public keys provided by Oathkeeper.
+func keyForToken(token *jwt.Token) (interface{}, error) {
+	if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+		return nil, errors.New(fmt.Sprintf("unexpected signing method: %v", token.Header["alg"]))
+	}
+	keySet, err := oathkeeperPublicKeys()
+	if err != nil {
+		return nil, err
+	}
+	// Should this be replaced by a deferred recover in case the value is not
+	// string and we panic? It might be faster than calling reflect every time.
+	if reflect.ValueOf(token.Header["kid"]).Kind() != reflect.String {
+		return nil, errors.New("invalid jwk header - the kid field is not a string")
+	}
+	keys := keySet.LookupKeyID(token.Header["kid"].(string))
+	if len(keys) == 0 {
+		return nil, errors.New("no suitable keys found")
+	}
+	return keys[0].Materialize()
+}
+
+// oathkeeperPublicKeys is a helper function that checks whether we have the
+// needed public key cached and if we don't it fetches it and caches it for us.
+//
+// See https://tools.ietf.org/html/rfc7517
+// See https://auth0.com/blog/navigating-rs256-and-jwks/
+// See http://self-issued.info/docs/draft-ietf-oauth-json-web-token.html
+// Encoding RSA pub key: https://play.golang.org/p/mLpOxS-5Fy
+func oathkeeperPublicKeys() (*jwk.Set, error) {
+	if oathkeeperPubKeys == nil {
+		r, err := http.Get(oathkeeperPubKeyUrl)
+		if err != nil {
+		    return nil, err
+		}
+		defer r.Body.Close()
+		b, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+		    return nil, err
+		}
+		set, err := jwk.ParseString(string(b))
+		if err != nil {
+			return nil, err
+		}
+		oathkeeperPubKeys = set
+	}
+	return oathkeeperPubKeys, nil
+}
+
+// tokenFromRequest extracts the JWT token from the request and returns it.
+// Returns an empty string if there is no token.
+func tokenFromRequest(r *http.Request) string {
```

Comment:
this is slightly better:

```
authHeader := r.Header.Get("Authorization")
parts := strings.Split(authHeader, "Bearer")
if len(parts) != 2 {
    // return error
}

return strings.TrimSpace(parts[1])
```

I think error handling on the header value is definitely nice to have, we can be more informative that way.
Without it I believe an invalid token, and an improperly formatted Authorization header are treated the same.